File: apps/web/src/types/external-integrations.ts
export type ExternalIntegrationCatalogField = {
  key: string;
  label: string;
  required: boolean;
  secret: boolean;
  placeholder?: string | null;
};

export type ExternalIntegrationCatalogItem = {
  integration_key: string;
  provider_key: string;
  display_name: string;
  category: string;
  adapter_type: string;
  description: string;
  sync_supported: boolean;
  manual_ingest_supported: boolean;
  consent_required: boolean;
  enabled: boolean;
  config_fields: ExternalIntegrationCatalogField[];
};

export type ExternalIntegrationConnection = {
  id: string;
  user_id: string;
  organization_id?: string | null;
  integration_key: string;
  provider_key: string;
  category: string;
  connection_status: string;
  consent_status: string;
  access_scope_json?: Record<string, unknown> | null;
  config_json?: Record<string, unknown> | null;
  metadata_json?: Record<string, unknown> | null;
  consented_at?: string | null;
  revoked_at?: string | null;
  last_synced_at?: string | null;
  last_sync_status?: string | null;
  last_error?: string | null;
  created_at: string;
  updated_at: string;
};

export type ExternalSyncJob = {
  id: string;
  connection_id: string;
  user_id: string;
  organization_id?: string | null;
  integration_key: string;
  provider_key: string;
  job_type: string;
  status: string;
  requested_by?: string | null;
  scheduled_at?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  sync_window_start?: string | null;
  sync_window_end?: string | null;
  signal_count: number;
  cursor_json?: Record<string, unknown> | null;
  request_payload_json?: Record<string, unknown> | null;
  result_payload_json?: Record<string, unknown> | null;
  error_message?: string | null;
  created_at: string;
  updated_at: string;
};

export type ExternalSignal = {
  id: string;
  connection_id: string;
  user_id: string;
  organization_id?: string | null;
  integration_key: string;
  provider_key: string;
  signal_type: string;
  source_item_id: string;
  signal_at?: string | null;
  signal_start_at?: string | null;
  signal_end_at?: string | null;
  numeric_value?: number | null;
  text_value?: string | null;
  unit?: string | null;
  signal_payload_json?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type UserExternalIntegrationStatus = {
  user_id: string;
  organization_id?: string | null;
  boundary_policy: Record<string, unknown>;
  catalog: ExternalIntegrationCatalogItem[];
  connections: ExternalIntegrationConnection[];
  recent_sync_jobs: ExternalSyncJob[];
  recent_signals: ExternalSignal[];
  signal_breakdown_by_type: Record<string, number>;
};

export type AdminExternalIntegrationOverview = {
  deployment_name: string;
  organization_id?: string | null;
  boundary_policy: Record<string, unknown>;
  catalog: ExternalIntegrationCatalogItem[];
  total_connections: number;
  active_connections: number;
  consented_connections: number;
  total_sync_jobs: number;
  queued_sync_jobs: number;
  failed_sync_jobs: number;
  total_signals: number;
  connection_breakdown_by_provider: Record<string, number>;
  signal_breakdown_by_type: Record<string, number>;
  recent_sync_jobs: ExternalSyncJob[];
};

export type ExternalIntegrationConnectionUpsertPayload = {
  integration_key: string;
  provider_key: string;
  consent_status?: string;
  access_scope_json?: Record<string, unknown> | null;
  config_json?: Record<string, unknown> | null;
  metadata_json?: Record<string, unknown> | null;
};

export type ExternalIntegrationSyncPayload = {
  job_type?: string;
  sync_window_days?: number | null;
  request_payload_json?: Record<string, unknown> | null;
};

export type ExternalIntegrationIngestPayload = {
  source_label?: string;
  payload_json: Record<string, unknown>;
};

export type ExternalIntegrationIngestResponse = {
  connection: ExternalIntegrationConnection;
  job: ExternalSyncJob;
  normalized_signal_count: number;
  signal_type_breakdown: Record<string, number>;
};

File: apps/api/src/mental_wellbeing_api/schemas/external_integrations.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class ExternalIntegrationCatalogFieldResponse(BaseModel):
    key: str
    label: str
    required: bool = False
    secret: bool = False
    placeholder: str | None = None


class ExternalIntegrationCatalogItemResponse(BaseModel):
    integration_key: str
    provider_key: str
    display_name: str
    category: str
    adapter_type: str
    description: str
    sync_supported: bool = True
    manual_ingest_supported: bool = True
    consent_required: bool = True
    enabled: bool = True
    config_fields: list[ExternalIntegrationCatalogFieldResponse] = Field(default_factory=list)


class ExternalIntegrationConnectionResponse(BaseModel):
    id: str
    user_id: str
    organization_id: str | None
    integration_key: str
    provider_key: str
    category: str
    connection_status: str
    consent_status: str
    access_scope_json: dict | None = None
    config_json: dict | None = None
    metadata_json: dict | None = None
    consented_at: datetime | None = None
    revoked_at: datetime | None = None
    last_synced_at: datetime | None = None
    last_sync_status: str | None = None
    last_error: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ExternalSyncJobResponse(BaseModel):
    id: str
    connection_id: str
    user_id: str
    organization_id: str | None
    integration_key: str
    provider_key: str
    job_type: str
    status: str
    requested_by: str | None = None
    scheduled_at: datetime | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    sync_window_start: datetime | None = None
    sync_window_end: datetime | None = None
    signal_count: int = 0
    cursor_json: dict | None = None
    request_payload_json: dict | None = None
    result_payload_json: dict | None = None
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ExternalSignalResponse(BaseModel):
    id: str
    connection_id: str
    user_id: str
    organization_id: str | None
    integration_key: str
    provider_key: str
    signal_type: str
    source_item_id: str
    signal_at: datetime | None = None
    signal_start_at: datetime | None = None
    signal_end_at: datetime | None = None
    numeric_value: float | None = None
    text_value: str | None = None
    unit: str | None = None
    signal_payload_json: dict | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ExternalIntegrationConnectionUpsertRequest(BaseModel):
    integration_key: str
    provider_key: str
    consent_status: str = "active"
    access_scope_json: dict[str, Any] | None = None
    config_json: dict[str, Any] | None = None
    metadata_json: dict[str, Any] | None = None


class ExternalIntegrationSyncRequest(BaseModel):
    job_type: str = "manual_sync"
    sync_window_days: int | None = Field(default=7, ge=1, le=90)
    request_payload_json: dict[str, Any] | None = None


class ExternalIntegrationIngestRequest(BaseModel):
    source_label: str = "manual_ingest"
    payload_json: dict[str, Any]


class ExternalIntegrationIngestResponse(BaseModel):
    connection: ExternalIntegrationConnectionResponse
    job: ExternalSyncJobResponse
    normalized_signal_count: int
    signal_type_breakdown: dict[str, int] = Field(default_factory=dict)


class UserExternalIntegrationStatusResponse(BaseModel):
    user_id: str
    organization_id: str | None = None
    boundary_policy: dict[str, Any] = Field(default_factory=dict)
    catalog: list[ExternalIntegrationCatalogItemResponse] = Field(default_factory=list)
    connections: list[ExternalIntegrationConnectionResponse] = Field(default_factory=list)
    recent_sync_jobs: list[ExternalSyncJobResponse] = Field(default_factory=list)
    recent_signals: list[ExternalSignalResponse] = Field(default_factory=list)
    signal_breakdown_by_type: dict[str, int] = Field(default_factory=dict)


class AdminExternalIntegrationOverviewResponse(BaseModel):
    deployment_name: str
    organization_id: str | None = None
    boundary_policy: dict[str, Any] = Field(default_factory=dict)
    catalog: list[ExternalIntegrationCatalogItemResponse] = Field(default_factory=list)
    total_connections: int = 0
    active_connections: int = 0
    consented_connections: int = 0
    total_sync_jobs: int = 0
    queued_sync_jobs: int = 0
    failed_sync_jobs: int = 0
    total_signals: int = 0
    connection_breakdown_by_provider: dict[str, int] = Field(default_factory=dict)
    signal_breakdown_by_type: dict[str, int] = Field(default_factory=dict)
    recent_sync_jobs: list[ExternalSyncJobResponse] = Field(default_factory=list)